models/deepseek/v4/sparse_attn.py

```
@@ -216,35 +221,102 @@ def sparse_attn(
+                pl.col_expand_mul(even_tile, cos_tile),
+                pl.col_expand_mul(odd_tile, sin_tile),
+            )
+            rope_odd_acc = pl.sub(
+                pl.col_expand_mul(odd_tile, cos_tile),
+                pl.col_expand_mul(even_tile, sin_tile),
+            )
+            rope_even = pl.assemble(rope_even, pl.cast(rope_even_acc, target_type=pl.BF16), [rope_head_row, 0])
+            rope_odd = pl.assemble(rope_odd, pl.cast(rope_odd_acc, target_type=pl.BF16), [rope_head_row, 0])
+
+        with pl.at(level=pl.Level.CORE_GROUP, name_hint="cfa_proj_rope_assemble"):
+            for r0 in pl.range(ROPE_CHUNK, HALF_ROPE, ROPE_CHUNK):
```

Ruff (0.15.12)

[warning] Loop control variable `r0` not used within loop body

(B007)